api/cpp/esp-idf/slint/src/esp_slint.cpp

```diff
@@ -0,0 +1,159 @@
+                        *touch_handle, touchpad_x, touchpad_y, NULL, &touchpad_cnt, 1);
+
+                if (touchpad_pressed && touchpad_cnt > 0) {
+                    // ESP_LOGI(TAG, "x: %i, y: %i", touchpad_x[0], touchpad_y[0]);
+                    last_touch_x = touchpad_x[0];
+                    last_touch_y = touchpad_y[0];
+                    m_window->window().dispatch_pointer_move_event(
+                            slint::LogicalPosition({ float(last_touch_x), float(last_touch_y) }));
+                    if (!touch_down) {
+                        m_window->window().dispatch_pointer_press_event(
+                                slint::LogicalPosition(
+                                        { float(last_touch_x), float(last_touch_y) }),
+                                slint::PointerEventButton::Left);
+                    }
+                    touch_down = true;
+                } else if (touch_down) {
+                    m_window->window().dispatch_pointer_release_event(
+                            slint::LogicalPosition({ float(last_touch_x), float(last_touch_y) }),
+                            slint::PointerEventButton::Left);
+                    m_window->window().dispatch_pointer_exit_event();
+                    touch_down = false;
+                }
+            }
+
+            if (std::exchange(m_window->needs_redraw, false)) {
+                auto region = m_window->m_renderer.render(buffer1, size.width);
+                auto o = region.bounding_box_origin();
+                auto s = region.bounding_box_size();
+                if (s.width > 0 && s.height > 0) {
+                    for (int y = o.y; y < o.y + s.height; y++) {
+                        for (int x = o.x; x < o.x + s.width; x++) {
+                            // Swap endianess to big endian
+                            auto px = reinterpret_cast<uint16_t *>(&buffer1[y * size.width + x]);
+                            *px = (*px << 8) | (*px >> 8);
+                        }
+                        esp_lcd_panel_draw_bitmap(panel_handle, o.x, y, o.x + s.width, y + 1,
+                                                  buffer1.data() + y * size.width + o.x);
+                    }
+                    if (buffer2) {
+                        std::swap(buffer1, buffer2.value());
+                    }
+                }
+            }
+
+            has_animations = m_window->window().has_active_animations();
+        }
+
+        TickType_t ticks_to_wait;
+        if (has_animations) {
+            continue;
+        } else {
+            if (auto wait_time = slint::platform::duration_until_next_timer_update()) {
+                ticks_to_wait = pdMS_TO_TICKS(wait_time->count());
+            } else {
+                ticks_to_wait = portMAX_DELAY;
+            }
+        }
+
+        // Poll at least every 30 ms for touch input. That's what LVGL does, too.
+        ticks_to_wait = std::min(ticks_to_wait, pdMS_TO_TICKS(30));
```

Comment:
I wonder if there is a better way

Reply:
On some devices there might be an interrupt line that could be used. But short of us having a way to test it or somebody contributing it, I think that's the way it has to be for now :(.

It might be worth making the existence of touch optional though, i.e. for devices without a touch controller.